Emoji: First pass at support in Interactions

obenland opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Fixes #970.

See https://mas.to/@obietester/115741669531612641
See https://obietester.blog/2025/12/02/battery-powered-tools-vs-air-compressor-tools-a-comprehensive-comparison/#comment-1946

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Adds callback function that replaces custom emoji strings with their image representations when available.
* Adds unit test to cover the new function.
* Runs author name and comment content through the new callback on insert and update.

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to Mastodon, for example, and reply to a federated test post with a custom emoji. `:yikes:`, `:AngeryCat:`, etc.
* Go to the comment list in wp-admin and make sure the emoji render as images.

obenland committed:
Emoji: First pass at support in Interactions

pfefferle commented:
I think before allowing images in comments, we have to think about/implement a proper blocking and moderation tooling (even more if it is about side-loading images): https://www.theverge.com/2023/7/24/23806093/mastodon-csam-study-decentralized-network

obenland committed:
Process emoji after sanitization

obenland commented:
I just pushed an update that moves the replacement into a filter that runs after comment_content and the author have been sanitized. That way, only custom emoji images will be added to the content.

jeherve commented:
> Heads up: This PR adds `img` tags to the list of allowed HTML tags for Interactions!

Would sideloading the image onto the site with `media_sideload_image` be helpful here?

- It would ensure the images remain available in the future.
- It would avoid image hotlinking, which can be frowned upon.
- It would add some validation to the image data before to simply display it on the site.

Of course, we'd need to be okay adding potentially a lot of new media to a site once that's enabled. I'm not sure that's okay.

obenland commented:
> It would avoid image hotlinking, which can be frowned upon.

With custom emoji being shared tags, is that not kind of expected?

jeherve commented:
> With custom emoji being shared tags, is that not kind of expected?

I'm honestly not sure what the common practice is with other Fediverse software. From what I can tell, on Mastodon and on GoToSocial the images from remote instances seem to be cached on the local server:

**GoToSocial:**

<img width="998" alt="image" src="https://github.com/user-attachments/assets/f659f2f7-4d4c-4a08-97a9-7b6d5a71922f" />

**Mastodon**

<img width="982" alt="image" src="https://github.com/user-attachments/assets/6946fe23-7f68-44db-9cb3-783a638ed1ca" />

obenland committed:
Save emoji images to media library

obenland committed:
Fix emoji display in wp-admin

Jiwoon-Kim commented:
## Introduction

To successfully introduce custom emoji functionality in WordPress, it is necessary to first standardize the way standard emojis are rendered across the WordPress ecosystem. Additionally, by analyzing the implementations of other Fediverse instances like Mastodon and Misskey, we can identify their limitations and strengths, allowing for a more strategic and robust approach to custom emoji support in WordPress.

---

## 1. The Need for Standardization in WordPress Emoji Rendering

Currently, the WordPress ecosystem features at least four different methods for rendering standard emojis, which creates fragmentation that must be addressed before introducing custom emojis.

- **Problem: Fragmented Rendering Methods**
  - **Mastodon/Misskey**: Render emojis using SVG assets hosted on their own servers.
  - **WordPress (Default)**: Uses the operating system’s system font.
  - **WordPress (Fallback)**: If the system font does not support certain emojis (e.g., flag emojis on Windows), the rendering method varies by environment.
    - **WordPress.com Blogs**: Use Twemoji SVGs from `s0.wp.com`.
    - **Self-hosted WordPress**: Use core emoji SVGs from `s.w.org`.
    - **WordPress Reader (for self-hosted posts)**: Emojis are displayed as PNG images proxied through `i0.wp.com`.

- **Improvement Direction**
  - Before introducing a custom emoji system, **WordPress.com, self-hosted WordPress, and the WordPress Reader should all render standard emojis in a consistent manner**. This requires a unified approach to emoji rendering across the ecosystem.

---

## 2. Analysis of Other Instances and WordPress Improvement Directions

Limitations found in other platforms provide actionable improvement goals for WordPress, while their strengths offer models to emulate.

### Mastodon’s Limitations: Inconsistent CDN Paths

- **Problem**: Mastodon uses different CDN paths for custom emojis displayed in user profiles versus posts. Even when viewing emojis from external instances, the cached paths vary, leading to inconsistency.
- **WordPress Improvement Direction**:
  - Learn from Mastodon’s example: WordPress should provide custom emojis through **a consistent path or proxy system**, regardless of where they are displayed (profiles, posts, comments).

### Misskey’s Strengths and Limitations

- **Strengths to Emulate**:
  - **Consistent Proxy Usage**: Misskey uses proxies like `proxy.misskeyusercontent.jp` to serve custom emojis, ensuring consistent paths across all contexts (profiles, notes, etc.).
  - **Rich Metadata**: Misskey supports extensive metadata for custom emojis (name, tags, category, NSFW status, local-only flag, license, etc.), enabling advanced management.
- **Limitations to Address**:
  - **Reaction Aggregation**: Even if the same custom emoji is used for reactions on different instances, Misskey (and Mastodon) cannot aggregate reactions across instances due to different original paths.
  - **WordPress Improvement Direction**:
    - Move beyond simply comparing image paths. **Design a mechanism to identify and aggregate reactions for identical custom emojis across instances**, possibly by using original source URLs or unique identifiers for each emoji[8].

---

## Key Considerations for Custom Emoji Implementation in WordPress

### 1️⃣ Unified Emoji Rendering

WordPress currently relies on system fonts for emoji display, but for unsupported emojis (like flags on Windows), it falls back to Twemoji SVGs from `s.w.org` or `s0.wp.com`. The fact that WordPress.com and self-hosted sites use different paths leads to compatibility issues, especially with the WordPress Reader.

**Solution:**
- **Set up a dedicated CDN path for custom emojis** (e.g., `custom-emoji.my-site.com/emoji/`).
- **Integrate Twemoji and custom emojis into a unified custom renderer** (e.g., using `Twemoji.parse()` in JavaScript).
- **Offer an option to disable system font fallback and enforce SVG/PNG rendering**.

### 2️⃣ Custom Emoji Registration and Management System

Mastodon and Misskey allow admins to add custom emojis via their admin panels, managed as JSON databases. WordPress should implement similar functionality via a plugin.

**Essential Features:**
- **Name** (`:emoji_name:`)
- **Tags and Categories**
- **NSFW Flag**
- **CDN Path**
- **License Information**
- **Local/Global Option** (whether to include emojis in Fediverse transmission)

### 3️⃣ Handling Custom Emojis in ActivityPub Integration

ActivityPub protocol can include emoji metadata. Mastodon and Misskey use the `customEmoji` field to declare each emoji’s path.

**Problem:**
- **Emoji reactions from different instances are not aggregated** because the same emoji name can have different paths on each instance.

**Improvement:**
- **Introduce a custom emoji federation registry** (e.g., `fediemoji.org`).
- **Identify emojis by unique ID or hash value**, with CDN path as optional.

### 4️⃣ Reader and External Embed Support

For custom emojis to display correctly in the WordPress Reader or via oEmbed, **CDN CORS must be enabled and the public path must be fixed**.

---

## Mastodon / Misskey Limitations and Improvement Possibilities

| Issue                                 | Current Approach                        | Improvement Direction                                  |
|----------------------------------------|-----------------------------------------|--------------------------------------------------------|
| **Emoji reaction aggregation fails**   | Same emoji name, different instance path| Common registry; identify by hash or GUID              |
| **Note vs. profile CDN path mismatch** | Mastodon: different cache paths         | Unify CDN cache paths or normalize paths               |
| **Lack of metadata (Mastodon)**        | Only name is registered                 | Add metadata like Misskey (tags, license, NSFW, etc.)  |
| **Embed/proxy path confusion**         | Misskey: proxy.misskeyusercontent.jp    | Proxy maintains original path; verify with CDN hash    |

---

## WordPress Implementation Strategy (Proposal)

### 📦 Custom Emoji Plugin Design

- **Twemoji-based custom parser**
- **Emoji metadata JSON management**
- **Register custom-emoji CDN path**
- **Integrate with ActivityPub plugin to inject `customEmoji` field**
- **Optional server-side reaction integration**

### 🌐 Custom Emoji Federation Registry API Design

- **Provide name, URL, hash, license, and category**
- **Allow instances to sync this list to their local DB**
- **Include original registry ID in ActivityPub transmissions**

---

## Conclusion

Custom emoji implementation is more than just image uploads.

- **Emoji identification** (name, hash)
- **Unified CDN path**
- **Standardization for Fediverse transmission**

These must be addressed in parallel for seamless federation. Mastodon and Misskey have encountered issues due to the lack of a unified approach. WordPress could take a leading role by launching a **custom emoji registry API** to solve these problems proactively[8].

---

## Summary Table: Emoji Rendering and Custom Emoji Management

| Platform            | Emoji Rendering Method                | Custom Emoji Management         | Key Issues/Strengths                |
|---------------------|---------------------------------------|---------------------------------|-------------------------------------|
| WordPress           | System font, Twemoji SVG, PNG proxy   | Limited, no standard plugin     | Fragmented, Reader compatibility    |
| Mastodon            | Server-hosted SVG, PNG                | Admin panel, JSON DB            | Inconsistent CDN paths              |
| Misskey             | Proxy, server-hosted SVG/PNG          | Admin panel, rich metadata      | Consistent proxy, rich metadata     |

---

By standardizing emoji rendering, introducing a robust custom emoji management system, and addressing federation challenges, WordPress can provide a seamless and expressive custom emoji experience for users, admins, and the wider Fediverse.

---
refer
제가 간략하게 알아본 바에 따르면, 마스토돈과 미스키의 경우 각 인스턴스에서 이모티콘을 렌더링합니다.
🥳 :partying_face:
https://mastodon.social/emoji/1f973.svg
https://uri.life/emoji/1f973.svg
https://misskey.io/twemoji/1f973.svg
워드프레스의 경우 시스템 폰트를 사용합니다.

문제는 시스템폰트가 없을 때 워드프레스가 이모지를 렌더링하는 방식입니다.
윈도우는 국기 이모지를 지원하지 않습니다.
🇰🇷
https://mastodon.social/emoji/1f1f0-1f1f7.svg
https://misskey.io/twemoji/1f1f0-1f1f7.svg

워드프레스닷컴의 블로그의 경우
https://s0.wp.com/wp-content/mu-plugins/wpcom-smileys/twemoji/2/svg/1f1f0-1f1f7.svg
워드프레스 리더에서 확인하면
🇰🇷
셀프호스팅 워드프레스의 경우
https://s.w.org/images/core/emoji/15.1.0/svg/1f1f0-1f1f7.svg
워드프레스 리더에서 확인하면
https://i0.wp.com/s.w.org/images/core/emoji/15.1.0/72x72/1f1f0-1f1f7.png?ssl=1

그러니까 워드프레스 생태계에서 이모지 렌더링 방식이 4가지나 되므로, 커스텀이모지까지 고려하게 되면 상당히 복잡해집니다.
어느정도 교통정리를 해아합니다.  

이제 커스텀이모지로 들어가봅시다.
mastodon.social의 :mastodon: 이모지
https://files.mastodon.social/custom_emojis/images/001/099/845/original/629dc18288be6387.png

uri.life에서 mastodon.social의 :mastodon: 이모지 확인
https://uri.life/@thaumiel999@mastodon.social/114651985535580673
https://life.uricdn.one/cache/custom_emojis/images/000/000/952/original/4e6b8b95bf17eee8.png

https://uri.life/@thaumiel999 (마스토돈 플랫폼)에서 :mastodon: 게시
https://life.uricdn.one/custom_emojis/images/000/001/376/original/d9a3907f06edd0cc.png
mastodon.social에서 확인
https://mastodon.social/@thaumiel999@uri.life/114651968583821676 
https://files.mastodon.social/cache/custom_emojis/images/000/010/391/original/ee5885e3bd7ffad8.png
https://uri.life/@thaumiel999@mastodon.social 프로필의 :mastodon: 아이콘 확인
https://life.uricdn.one/cache/custom_emojis/images/000/000/952/original/4e6b8b95bf17eee8.png

마스토돈에서는 프로필과 노트의 커스텀 이모지 cdn이 다르다는 것을 확인할 수 있음. 개선필요

미스키에서 mastodon.social의 :mastodon: 이모지 확인
노트/프로필
https://proxy.misskeyusercontent.jp/image/files.mastodon.social%2Fcustom_emojis%2Fimages%2F001%2F099%2F845%2Foriginal%2F629dc18288be6387.png?emoji=1
인스턴스 파비콘 미리보기
https://proxy.misskeyusercontent.jp/preview/mastodon.social%2Fpacks%2Fassets%2Ffavicon-48x48-DMnduFKh.png?fallback=1&preview=1

misskey.io의 :misskey: 커스텀 이모지
프로필과 노트에서 일관된 경로
https://proxy.misskeyusercontent.jp/image/media.misskeyusercontent.jp%2Femoji%2Fmisskey.png?emoji=1
마스토돈에서 미스키 이모지 확인
프로필과 노트 모두에서 일관된 경로
https://files.mastodon.social/cache/custom_emojis/images/000/273/113/original/b29393e466430baf.png

https://daepi.so/proxy/image.webp?url=https%3A%2F%2Fmedia.misskeyusercontent.com%2Femoji%2Fmisskey.png&emoji=1

미스키는 이모지 정보에 이름, 태그, 카테고리, 열람주의, 로컬에만, 라이선스 등 옵션이 있음.
https://misskey.io/notes/a8sajiih9ywy0851
https://proxy.misskeyusercontent.jp/image/cdn.daepi.so%2Fimages%2F35e15e31-241d-4ef8-8ae4-87173daffef9?emoji=1
https://proxy.misskeyusercontent.jp/image/media.misskeyusercontent.jp%2Femoji%2Fmisskey.png?emoji=1
특이한점으로는 같은 커스텀이모지 :misskey: 인데 인스턴스가 다르면 이모지 리액션이 합산이 안됨. 오리지널 경로가 아닌 다른 인스턴스 이모지 리액션이 안보임. 개선 필요.

obenland commented:
@Jiwoon-Kim I appreciate your input and feedback, but comments and issues of this length are not helpful. They generally lack a specific ask or suggestion and are incredibly hard to read. Going forward, please keep comments/issues concise and actionable, like [I asked for previously](https://github.com/Automattic/wordpress-activitypub/issues/1734#issuecomment-2913538223).

Jiwoon-Kim commented:
@obenland  Thank you for your feedback, and I really appreciate the attention.

As someone coming from a non-developer background, I tend to approach these topics from a more experiential and design-oriented perspective. That often leads me to start from broader conceptual ideas before narrowing down to specifics—especially when things are interconnected or have potential design implications that aren't immediately obvious.

I understand that this approach can result in comments that feel messy or overwhelming, and I apologize if it made the discussion harder to follow. I added the reference materials at the end to reduce the need for re-investigation, but I realize now that might have come across as overly verbose.

Also, since this pull request already exists, I thought discussing the context and UX considerations here would help avoid scattering the conversation into too many isolated issues. But I see now that clarity and conciseness in individual issues is important, so I’ll aim to keep future contributions more structured and actionable.

If you ever have time, I’d be grateful for any advice on how to break down larger conceptual suggestions into well-scoped issues. I’m eager to learn and collaborate more effectively with the team.

Thanks again!

---

Also, just to explain a bit about my workflow — I usually start with an idea in Korean, then use GPT to help flesh it out and translate it into English. Sometimes I’ll use Gemini or Perplexity to refine the wording too. So I’m not actually writing or editing in English directly most of the time.

If this were a Korean-language project, breaking things down into smaller, well-scoped issues like you suggested would honestly be much easier for me. But when working in English, it takes quite a lot more energy for me to carefully read through, translate, and restructure everything. That’s why I sometimes rely on machine translation to double-check and just post the draft as it is.

Hope you can understand that part of my process.

Anyway — I actually ran your feedback and my original comment through Gemini to auto-organize it a bit, and I’ll drop that summary in a follow-up comment below.

obenland commented:
I do understand, maybe part of your workflow could be to instruct your LLM to use natural language and distill its response to a sentence or two?

Jiwoon-Kim commented:
@obenland 
Thank you for your understanding—I really appreciate it.

I agree that breaking down larger ideas into focused issues is a good approach, and I’ll try to normalize my prompts for LLMs to help with this in the future. However, it can be challenging to distill complex or interconnected ideas into just a sentence or two, especially when I’m trying to preserve the overall context and implications.

Here’s a bit more about my current workflow:  
When I have an idea, I usually start by brainstorming in Korean and use GPT to help structure and clarify my thoughts. GPT is great at reorganizing fragmented concepts into coherent language and suggesting implementation directions. If the initial explanation is lacking, I add more context to make sure the AI understands the full picture. This process often involves several iterations and a fair amount of trial and error.

Once I’m satisfied with the structure, I use PowerToys’ Advanced Paste feature to organize everything into a markdown text file. I then use AI translation to convert it into English. Even just getting to this stage takes a lot of time and mental effort before I’m ready to post an issue on GitHub.

There’s a reason I sometimes prefer to bundle ideas together:  
Gemini, for example, works best with very specific, focused requests. It excels at quickly categorizing and summarizing concise, scannable content where the desired outcome is clear. However, if I split every idea into separate issues, the overall context and the “big picture” of my suggestion can get lost. Additionally, there are limits to how many requests I can make per day, so I try to use them carefully.

I’ll definitely take your advice and experiment with using LLMs to summarize and distill my suggestions more effectively. Thanks again for your patience and feedback—I’m eager to keep improving my workflow and collaborating better with the team!

---

Below is the process and set of prompts I plan to use moving forward. I’m sharing them here as they may be useful for others as well.

---

## 1. Breaking Down a Long Proposal into Focused GitHub Issues

**Prompt:**
> Act as a senior software project maintainer. Your task is to review the following long-form proposal and break it down into a list of smaller, focused, and actionable GitHub issues.
>
> For each proposed issue, provide a concise title and a one-sentence summary describing its core request. Ensure that each issue is independent enough to be discussed and implemented separately.
>
> Here is the proposal:
> [Paste the entire long-form proposal here, either in Korean or as a first draft organized by AI.]

---

## 2. Crafting a Concise and Actionable GitHub Issue

**Prompt:**
> Using the context provided below, generate a concise and actionable GitHub issue in English. Structure the output in a standard markdown format with the following sections:
>
> - **### Is your feature request related to a problem? Please describe(Problem statement)
> - **### Describe the solutionou'd like (Desired solution)
> - **### Describe alternatives you've consideredConsidered alternatives)
>
> The tone should be professional, clear, and suitable for an open-source project.
>
> Here is the context for the issue titled "[Title generated in Step 1]":
> [Paste only the content relevant to this issue here.]

---

## 3. Summarizing an Issue or Comment for Quick Triage

**Prompt:**
> Distill the following GitHub issue/comment into its most essential, scannable points. The output should be a short bulleted list or 1-2 concise sentences that clearly state the core problem and the proposed action. The goal is to make it easy for a maintainer to triage the issue quickly.
>
> Here is the text to distill:
> [Paste the issue content or long comment generated in Step 2 here.]

---

Feel free to adapt or use these prompts as needed!

Jiwoon-Kim commented:
https://developer.wordpress.org/news/2025/08/registering-custom-social-icons-in-wordpress-6-9/

- https://www.w3.org/wiki/ActivityPub/Primer/Hashtags
- https://socialhub.activitypub.rocks/t/fep-9098-custom-emojis/5408/6
- https://codeberg.org/fediverse/fep/src/branch/main/fep/9098/fep-9098.md
- https://misskey-hub.net/ns/

Jiwoon-Kim commented:
- https://misskey.io/about#emojis
- https://[misskey.io/twemoji/1f973.svg](https://misskey.io/twemoji/1f973.svg)
- https://[mastodon.social/emoji/1f1f0-1f1f7.svg](https://mastodon.social/emoji/1f1f0-1f1f7.svg)
- https://misskey-hub.net/en/docs/for-users/resources/glossary/#custom-emoji
- https://misskey-hub.net/en/docs/for-developers/api/endpoints/
- https://misskey.io/api-doc
- https://misskey.io/api-doc#operation/admin___emoji___import-zip
- https://misskey-hub.net/en/docs/for-developers/api/permission/

obenland committed:
Optimize emoji processing performance

Replace multiple database queries with single batched query and fast lookup arrays. This eliminates N+1 query problem where each emoji required separate database calls.

Changes:
- Single get_posts() query instead of one per emoji
- Fast O(1) lookup arrays instead of O(n) loops
- Consolidated meta query building into single loop
- Removed redundant array collections

obenland committed:
Fix temp file cleanup for emoji downloads

Move temp file cleanup inside success block to prevent cleanup attempts when no file was created. This ensures temp files are properly cleaned up only when download_url() succeeds.

obenland committed:
Set timeout for emoji downloads

Adds a 10 second timeout to the download_url call when downloading emoji files to prevent long-running requests.

obenland committed:
Fix fragile emoji detection in comment author names

- Use get_comment_author filter instead of comment_author for proper timing
- Improve emoji detection to check for class="emoji" instead of just "emoji"
- Use proper HTML entity decoding with ENT_QUOTES | ENT_HTML5 flags
- Add test coverage for emoji in comment author names

obenland committed:
Refactor emoji processing methods for better maintainability

Break up large replace_custom_emoji method into focused single-responsibility methods:
- extract_emoji_data(): Parse emoji from activity tags
- get_emoji_attachments(): Handle database queries and build lookup arrays
- get_or_create_emoji_attachment(): Decision logic for reuse vs download
- download_emoji(): Handle file download and WordPress attachment creation
- replace_emoji_in_text(): Handle text replacement with HTML

Improvements:
- Better separation of concerns following SRP
- Improved documentation with detailed parameter types
- Cleaner alt text (removes colons from emoji names)
- Simplified fallback logic using consistent emoji URLs

obenland committed:
Refactor emoji handling to dedicated Emoji class

Moved custom emoji processing logic from Interactions to a new Activitypub\Emoji class for better separation of concerns and maintainability. Updated Interactions and related tests to use the new Emoji class.

Jiwoon-Kim commented:
- https://codeberg.org/fediverse/fep/src/branch/main/fep/c0e0/fep-c0e0.md
- https://docs.akkoma.dev/stable/development/ap_extensions/#emoji-reactions

obenland committed:
Fix tests

obenland committed:
whitespace

obenland committed:
Store emoji shortcodes in database, replace on display

Instead of storing HTML markup in the comment_author field, now stores
the shortcode (e.g., ":emoji:") and replaces it on display. This makes
the implementation more future-proof:

- If the plugin is disabled, shortcodes display gracefully
- Emoji images can be updated dynamically
- Database stays clean without HTML markup

Implementation:
- Store ActivityPub emoji tag data in comment meta
- Use get_comment_author filter to replace shortcodes with img tags
- Use comment_author filter to selectively unescape only emoji images
- Maintain WordPress security by only unescaping known emoji markup

obenland commented:
Todo: 
* Store emoji files for comment authors on comment save.
* Update replacement with using attachment

Jiwoon-Kim commented:
It’s possible to display custom emojis from remote instances by extending the Smiley feature.
See: https://wordpress.org/documentation/article/what-are-smilies/#where-are-my-smiley-images-kept

https://github.com/Automattic/wordpress-activitypub/discussions/2371

obenland committed:
Remove merge conflict remnant

obenland committed:
Refactor emoji to use Attachments class for file storage

- Add import_emoji() method to Attachments class for downloading and caching emoji images
- Store emoji files organized by source domain: /activitypub/emoji/{domain}/
- Simplify Emoji class from 225 lines to 89 lines by delegating storage to Attachments
- Remove WordPress attachment posts for emoji, use lightweight file storage instead
- Add img tag to allowed HTML in comments for emoji support
- Fix comment_author to store shortcodes (replaced at display time, not save time)

obenland committed:
Centralize emoji handling in Emoji class

- Add prepare_comment_data() for content emoji replacement at insert-time
- Add prepare_actor_meta() for storing emoji data on remote actors
- Add replace_from_json() for display-time emoji replacement
- Store emoji on remote actors instead of duplicating in comment meta
- Look up author emoji from linked remote actor at display time
- Remove redundant pre_comment_content filters from Interactions

obenland committed:
Store only emoji tags in actor meta

- Add get_emoji_tags() to filter only Emoji-type tags
- Store filtered emoji tags instead of entire tag array
- Use to_json() and to_array() while filters are removed

obenland committed:
Enhance emoji support with accessibility and caching improvements

- Add accessibility attributes to emoji img tags (width, height, title, draggable)
- Add CSS for reduced-motion preference to pause animated emoji
- Update allowed HTML attributes for emoji in comments
- Add cache invalidation via filemtime for emoji with updated timestamps
- Add emoji support for actor names and descriptions in follower emails
- Add tests for emoji cache invalidation behavior

obenland committed:
Fix emoji method name and test assertions

- Use correct method Remote_Actors::get_by_uri() instead of get_by_actor()
- Update test assertions to check for locally cached emoji URLs

obenland committed:
Add migration to populate emoji meta for existing remote actors

Extracts emoji data from stored actor JSON and saves it as
_activitypub_emoji post meta, enabling display-time emoji
replacement for comments from actors that were stored before
emoji support was added.

obenland committed:
Fix case-insensitive emoji replacement and add unit tests

Mastodon sends emoji names in lowercase in tags but may use different
casing in the content (e.g., :KannaWave: vs :kannawave:). Changed
str_replace to str_ireplace for case-insensitive matching.

Added comprehensive unit tests for the Emoji class covering multiple
emoji, repeated emoji, adjacent emoji, and case-insensitive replacement.

obenland committed:
Move emoji CSS to inline style on comments screen

Use wp_add_inline_style to add emoji float fix only on edit-comments.php,
attached to wp-emoji-styles handle. Removes unused CSS from admin stylesheet.

obenland committed:
Merge branch 'trunk' into add/emoji-support

obenland commented:
@pfefferle Might need some testing, but should be close now.

See https://mas.to/@obietester/115741669531612641
See https://obietester.blog/2025/12/02/battery-powered-tools-vs-air-compressor-tools-a-comprehensive-comparison/#comment-1946

obenland committed:
Add changelog entry

obenland committed:
Merge branch 'add/emoji-support' of https://github.com/Automattic/wordpress-activitypub into add/emoji-support

obenland commented on includes/class-attachments.php:
```diff
@@ -159,6 +166,78 @@ public static function import_post_files( $attachments, $post_id ) {
+		// Initialize filesystem.
+		\WP_Filesystem();
+		global $wp_filesystem;
+
+		// Move file to destination (overwrite if exists).
+		if ( ! $wp_filesystem->move( $tmp_file, $file_path, true ) ) {
+			\wp_delete_file( $tmp_file );
```
Returning false is consistent with existing patterns.

obenland commented on includes/class-mailer.php:
```diff
@@ -172,6 +172,14 @@ public static function new_follower( $activity, $user_ids, $success ) {
 		$actor = self::normalize_actor( $actor );
 
+		// Replace emoji in actor name and summary.
+		if ( ! empty( $actor['name'] ) ) {
+			$actor['name'] = Emoji::replace_for_actor( $actor['name'], $actor['url'] );
```
Both classes are in the same `Activitypub` namespace - no use statement needed.

obenland commented on includes/class-emoji.php:
```diff
@@ -0,0 +1,202 @@
+	 *
+	 * @param array $comment_data The comment data array.
+	 * @param array $activity     The activity array.
+	 *
+	 * @return array The comment data with emoji processing applied.
+	 */
+	public static function prepare_comment_data( $comment_data, $activity ) {
+		// Replace emoji in content at insert-time.
+		if ( ! empty( $comment_data['comment_content'] ) && ! empty( $activity['object'] ) ) {
+			$comment_data['comment_content'] = self::replace_custom_emoji( $comment_data['comment_content'], $activity['object'] );
+		}
+
+		return $comment_data;
```
Intentional design: content uses insert-time replacement (permanent cache), author uses display-time (updates with actor profile). Already documented in docblock.

obenland commented on includes/class-emoji.php:
```diff
@@ -0,0 +1,202 @@
+	private static function replace_emoji_in_text( $text, $placeholder, $emoji_url ) {
+		$name = trim( $placeholder, ':' );
+
+		return str_ireplace(
+			$placeholder,
+			sprintf(
+				'<img src="%s" alt="%s" title="%s" class="emoji" width="20" height="20" draggable="false" />',
+				\esc_url( $emoji_url ),
+				\esc_attr( $name ),
+				\esc_attr( $name )
+			),
+			$text
+		);
```
Case-insensitive is intentional. Mastodon/other platforms vary in emoji name casing (:Kappa: vs :kappa:). Matching should be flexible.

obenland commented on includes/wp-admin/class-admin.php:
```diff
@@ -335,6 +335,13 @@ public static function enqueue_scripts( $hook_suffix ) {
 			);
 		}
+
+		if ( 'edit-comments.php' === $hook_suffix ) {
+			\wp_add_inline_style(
+				'wp-emoji-styles',
+				'.column-author img.emoji { float: none; }'
+			);
+		}
```
The CSS only needs to apply on the comments screen where emoji appear and images get floated.

obenland commented on includes/class-emoji.php:
```diff
@@ -0,0 +1,202 @@
+	private static function replace_emoji_in_text( $text, $placeholder, $emoji_url ) {
+		$name = trim( $placeholder, ':' );
+
+		return str_ireplace(
+			$placeholder,
+			sprintf(
+				'<img src="%s" alt="%s" title="%s" class="emoji" width="20" height="20" draggable="false" />',
+				\esc_url( $emoji_url ),
+				\esc_attr( $name ),
+				\esc_attr( $name )
```
Might be a good follow-up, this should be good for now.

obenland commented on includes/class-migration.php:
```diff
@@ -215,6 +216,10 @@ public static function maybe_migrate() {
 			\wp_schedule_single_event( \time(), 'activitypub_migrate_avatar_to_remote_actors' );
 		}
 
+		if ( \version_compare( $version_from_db, 'unreleased', '<' ) ) {
+			\wp_schedule_single_event( \time(), 'activitypub_migrate_actor_emoji' );
+		}
```
Will happen on release.

obenland committed:
Address PR review feedback for emoji support

- Validate emoji downloads are actual images using wp_get_image_mime()
- Handle wp_mkdir_p() failure by cleaning up temp file
- Use glob for timestamp comparison to handle webp-optimized files
- Add unslash/re-slash in prepare_comment_data() to avoid escaping img attributes
- Add test for download failure handling

obenland commented on includes/class-attachments.php:
```diff
@@ -159,6 +166,78 @@ public static function import_post_files( $attachments, $post_id ) {
+			$local_time  = \file_exists( $file_path ) ? \filemtime( $file_path ) : 0;
+			$remote_time = \strtotime( $updated );
+
+			if ( $remote_time && $local_time >= $remote_time ) {
+				return $cached_url;
+			}
+		}
+
+		if ( ! \function_exists( 'download_url' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/file.php';
+		}
+
+		// Download the emoji.
+		$tmp_file = \download_url( $emoji_url, 10 ); // 10 second timeout for emoji downloads.
+		if ( \is_wp_error( $tmp_file ) ) {
+			return false;
+		}
+
+		// Get storage paths for this emoji.
+		$paths = self::get_emoji_storage_paths( $emoji_url );
+
+		// Create directory if it doesn't exist.
+		\wp_mkdir_p( $paths['basedir'] );
+
+		// Generate filename from URL path (consistent with get_emoji_url lookup).
+		$url_path  = \wp_parse_url( $emoji_url, PHP_URL_PATH );
+		$file_name = \sanitize_file_name( \basename( $url_path ) );
+		$file_path = $paths['basedir'] . '/' . $file_name;
+
+		// Initialize filesystem.
+		\WP_Filesystem();
+		global $wp_filesystem;
+
+		// Move file to destination (overwrite if exists).
+		if ( ! $wp_filesystem->move( $tmp_file, $file_path, true ) ) {
+			\wp_delete_file( $tmp_file );
+			return false;
+		}
+
+		return $paths['baseurl'] . '/' . $file_name;
```
Fixed in [1817ec5](https://github.com/Automattic/wordpress-activitypub/pull/1129/commits/1817ec586814d7a55b1e4e2eef0cf90c58a01d88).

obenland commented on includes/class-attachments.php:
```diff
@@ -159,6 +166,78 @@ public static function import_post_files( $attachments, $post_id ) {
+		$paths = self::get_emoji_storage_paths( $emoji_url );
+
+		// Create directory if it doesn't exist.
+		\wp_mkdir_p( $paths['basedir'] );
```
Fixed in [1817ec5](https://github.com/Automattic/wordpress-activitypub/pull/1129/commits/1817ec586814d7a55b1e4e2eef0cf90c58a01d88).

obenland commented on tests/phpunit/tests/includes/class-test-attachments.php:
```diff
@@ -1204,4 +1204,104 @@ public function test_delete_actors_directory_ignores_non_actors() {
+		$this->assertTrue( $download_attempted, 'Should have attempted to re-download the emoji' );
+	}
+
+	/**
+	 * Test emoji import returns false for invalid URL.
+	 *
+	 * @covers ::import_emoji
+	 */
+	public function test_import_emoji_returns_false_for_invalid_url() {
+		$this->assertFalse( Attachments::import_emoji( '' ) );
+		$this->assertFalse( Attachments::import_emoji( 'not-a-url' ) );
+	}
```
Fixed in [1817ec5](https://github.com/Automattic/wordpress-activitypub/pull/1129/commits/1817ec586814d7a55b1e4e2eef0cf90c58a01d88).

obenland commented on includes/class-attachments.php:
```diff
@@ -159,6 +166,78 @@ public static function import_post_files( $attachments, $post_id ) {
+			$url_path    = \wp_parse_url( $emoji_url, PHP_URL_PATH );
+			$file_name   = \sanitize_file_name( \basename( $url_path ) );
+			$file_path   = $paths['basedir'] . '/' . $file_name;
+			$local_time  = \file_exists( $file_path ) ? \filemtime( $file_path ) : 0;
```
Fixed in [1817ec5](https://github.com/Automattic/wordpress-activitypub/pull/1129/commits/1817ec586814d7a55b1e4e2eef0cf90c58a01d88).

obenland commented on includes/collection/class-interactions.php:
```diff
@@ -109,6 +110,7 @@ public static function update_comment( $activity ) {
 		// Found a local comment id.
 		$comment_data['comment_author']  = \esc_attr( empty( $meta['name'] ) ? $meta['preferredUsername'] : $meta['name'] );
 		$comment_data['comment_content'] = \addslashes( $activity['object']['content'] );
+		$comment_data                    = Emoji::prepare_comment_data( $comment_data, $activity );
```
Fixed in [1817ec5](https://github.com/Automattic/wordpress-activitypub/pull/1129/commits/1817ec586814d7a55b1e4e2eef0cf90c58a01d88).

obenland commented on includes/class-attachments.php:
```diff
@@ -242,6 +321,65 @@ private static function get_storage_paths( $object_id, $object_type ) {
+		$url_name  = \pathinfo( $url_path, PATHINFO_FILENAME );
+		$file_name = \sanitize_file_name( $url_name );
+
+		// Look for file with any extension (original or webp after optimization).
+		$files = \glob( $paths['basedir'] . '/' . $file_name . '.*' );
+
+		if ( ! empty( $files ) ) {
+			return $paths['baseurl'] . '/' . \basename( $files[0] );
```
Fixed in [1817ec5](https://github.com/Automattic/wordpress-activitypub/pull/1129/commits/1817ec586814d7a55b1e4e2eef0cf90c58a01d88).

obenland commented on includes/class-migration.php:
```diff
@@ -1156,4 +1161,55 @@ public static function migrate_avatar_to_remote_actors( $batch_size = 50 ) {
+				'offset'         => $offset,
+				'post_status'    => 'any',
+				'orderby'        => 'ID',
+				'order'          => 'ASC',
```
Breaks offset-based batching pattern used by other migrations.

obenland commented on includes/class-emoji.php:
```diff
@@ -0,0 +1,202 @@
+			return array();
+		}
+
+		return array(
+			'_activitypub_emoji' => \wp_json_encode( $emoji_tags ),
```
Chances of it failing are quite low.

obenland commented on tests/phpunit/tests/includes/collection/class-test-interactions.php:
```diff
@@ -710,6 +714,117 @@ function () {
+					array(
+						'type' => 'Emoji',
+						'name' => ':smile:',
+						'icon' => array(
+							'type'      => 'Image',
+							'mediaType' => 'image/png',
+							'url'       => 'https://example.com/files/smile.png',
+						),
+					),
+				),
+			),
+		);
+
+		$comment_id = Interactions::add_comment( $activity );
+		$comment    = get_comment( $comment_id );
+
+		// Test emoji replacement in comment content (emoji is cached locally).
+		$this->assertStringContainsString( 'kappa.png" alt="kappa" title="kappa" class="emoji"', $comment->comment_content );
+		$this->assertStringContainsString( 'smile.png" alt="smile" title="smile" class="emoji"', $comment->comment_content );
+		$this->assertStringNotContainsString( ':kappa:', $comment->comment_content );
+		$this->assertStringNotContainsString( ':smile:', $comment->comment_content );
+
+		// Test that shortcode is stored in database, not HTML.
+		$this->assertSame( 'Test User :kappa:', $comment->comment_author );
+
+		// Test that comment is linked to remote actor.
+		$remote_actor_id = get_comment_meta( $comment_id, '_activitypub_remote_actor_id', true );
+		$this->assertEquals( $actor_post_id, $remote_actor_id );
+
+		// Test that emoji data is stored on remote actor.
+		$emoji_data = get_post_meta( $actor_post_id, '_activitypub_emoji', true );
+		$this->assertNotEmpty( $emoji_data );
+
+		// Test emoji replacement on display via comment_author filter.
+		$author_with_emoji = get_comment_author( $comment_id );
+		$this->assertStringContainsString( 'kappa.png" alt="kappa" title="kappa" class="emoji"', $author_with_emoji );
+		$this->assertStringNotContainsString( ':kappa:', $author_with_emoji );
+
+		\remove_filter( 'pre_get_remote_metadata_by_actor', $filter, 10 );
+	}
```
Both insert and update use the same `prepare_comment_data() ` function - existing tests should cover the logic.

obenland committed:
Harden emoji KSES to only allow local emoji images

- Add Emoji::get_kses_allowed_html() with strict validation:
  - Requires class="emoji" attribute
  - Validates src URL points to local emoji uploads directory
  - Requires standard emoji dimensions (20x20)
- Update Interactions::allowed_comment_html() to use strict KSES
- Update Comment::unescape_emoji() to use strict KSES
- Only replace emoji shortcodes when local import succeeds
- Add activitypub_pre_import_emoji filter for test mocking
- Non-emoji img tags are now stripped from incoming content

obenland commented on includes/class-comment.php:
```diff
@@ -841,4 +843,49 @@ public static function hide_for() {
+			$author = \html_entity_decode( $author, ENT_QUOTES | ENT_HTML5, 'UTF-8' );
+		}
+
+		return $author;
```
Fixed in [b2413bc](https://github.com/Automattic/wordpress-activitypub/pull/1129/commits/b2413bc2e3e0a7871bc30c19fd86e566f773f735).

obenland commented on includes/collection/class-interactions.php:
```diff
@@ -311,6 +313,19 @@ public static function allowed_comment_html( $allowed_tags, $context = '' ) {
+				'height'    => true,
+				'draggable' => true,
+			);
+		}
```
Fixed in [b2413bc](https://github.com/Automattic/wordpress-activitypub/pull/1129/commits/b2413bc2e3e0a7871bc30c19fd86e566f773f735).

obenland commented on includes/class-mailer.php:
```diff
@@ -172,6 +172,14 @@ public static function new_follower( $activity, $user_ids, $success ) {
 
 		$actor = self::normalize_actor( $actor );
 
+		// Replace emoji in actor name and summary.
+		if ( ! empty( $actor['name'] ) ) {
+			$actor['name'] = Emoji::replace_for_actor( $actor['name'], $actor['url'] );
+		}
+		if ( ! empty( $actor['summary'] ) ) {
+			$actor['summary'] = Emoji::replace_for_actor( $actor['summary'], $actor['url'] );
+		}
```
They're in the same namespace.

obenland commented on includes/class-migration.php:
```diff
@@ -1156,4 +1161,55 @@ public static function migrate_avatar_to_remote_actors( $batch_size = 50 ) {
+				continue;
+			}
+
+			$emoji_meta = Emoji::prepare_actor_meta( $actor_data );
```
Same namespace.

obenland committed:
Add tests for glob metacharacter sanitization in emoji URLs

Verifies that sanitize_file_name() prevents glob patterns like
[abc].png, test*.png, tes?.png, and {foo,bar}.png from matching
unintended cached files.

obenland commented on includes/class-attachments.php:
```diff
@@ -242,6 +345,65 @@ private static function get_storage_paths( $object_id, $object_type ) {
+		$file_name = \sanitize_file_name( $url_name );
+
+		// Look for file with any extension (original or webp after optimization).
+		$files = \glob( $paths['basedir'] . '/' . $file_name . '.*' );
```
`sanitize_file_name()` removes glob metacharacters (`[, ], *, ?, {, }`). Added unit tests in 7faae095 to verify this behavior with data provider.

obenland commented on includes/class-attachments.php:
```diff
@@ -159,6 +166,102 @@ public static function import_post_files( $attachments, $post_id ) {
+			}
+
+			// Compare timestamps - re-download if remote is newer.
+			$paths       = self::get_emoji_storage_paths( $emoji_url );
+			$url_path    = \wp_parse_url( $emoji_url, PHP_URL_PATH );
+			$file_stem   = \sanitize_file_name( \pathinfo( $url_path, PATHINFO_FILENAME ) );
+			$matches     = \glob( $paths['basedir'] . '/' . $file_stem . '.*' );
+			$file_path   = ( $matches && \is_file( $matches[0] ) ) ? $matches[0] : null;
+			$local_time  = $file_path ? \filemtime( $file_path ) : 0;
```
`sanitize_file_name()` removes glob metacharacters (`[, ], *, ?, {, }`). Added unit tests in 7faae095 to verify this behavior with data provider.

obenland commented on includes/class-emoji.php:
```diff
@@ -0,0 +1,263 @@
+	private static function replace_emoji_in_text( $text, $placeholder, $emoji_url ) {
+		$name = trim( $placeholder, ':' );
+
+		return str_ireplace(
+			$placeholder,
+			sprintf(
+				'<img src="%s" alt="%s" title="%s" class="emoji" width="20" height="20" draggable="false" />',
+				\esc_url( $emoji_url ),
+				\esc_attr( $name ),
+				\esc_attr( $name )
+			),
+			$text
+		);
```
The performance impact is negligible for typical emoji counts (1-5 per comment). Adding regex compilation or async processing would add complexity without meaningful benefit.

obenland commented on includes/class-attachments.php:
```diff
@@ -159,6 +166,102 @@ public static function import_post_files( $attachments, $post_id ) {
+		if ( ! \function_exists( 'download_url' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/file.php';
+		}
+
```
`download_url()` uses `wp_safe_remote_get()` internally, which validates URLs via `wp_http_validate_url()`.

obenland committed:
Fix test docblocks and add emoji import mock

- Add missing @param tags for data provider test
- Add emoji import mock to interactions test for KSES validation

obenland committed:
Merge branch 'trunk' into add/emoji-support

pfefferle committed:
Merge branch 'trunk' into add/emoji-support

pfefferle committed:
Add reference to FEP-9098: Custom Emojis

Updated FEDERATION.md to include FEP-9098 in the list of supported FEPs. Added a reference to FEP-9098 in the Emoji class docblock for better documentation and traceability.

pfefferle committed:
Add array check for tag processing in Emoji class

Added is_array checks before accessing tag properties in two methods to prevent errors when non-array values are present in the tag list.

pfefferle committed:
Merge branch 'trunk' into add/emoji-support

obenland commented:
w00t! 🎉
